app/Http/Controllers/Lti3Controller.php

```diff
@@ -57,35 +60,69 @@ public function launch(Request $request)
         $settings["canvas_user_id"] = (string)$settings['canvas_user_id'];
         $settings["canvas_course_id"] = (string)$settings['canvas_course_id'];
         $settings["canvas_user_login_id"] = $launch->get_launch_data()['email'];
+        try{
+            $settings = $this->get_categories($settings);
 
-        $settings_new = [];
+            } catch (GuzzleException $e)
+        {
+            throw new LtiException("Error at LTIv3 get categories from canvas :" . $e->getMessage());
+
+        }
 
+        $settings_new = [];
         foreach ($settings as $key => $value) {
             Arr::set($settings_new, 'settings.custom_' . $key, $value);
         }
         session(['settings' => $settings_new['settings']]);
 
         logger("Lti3Middleware has settings.");
 
-        $this->parse_token($launch);
-        if ($launch->is_resource_launch()) {
-            return view('lti.index');
-        } else if ($launch->is_deep_link_launch()) {
-            echo 'Deep Linking Launch!';
-        } else {
-            echo 'Unknown launch type';
-        }
         return view('lti.index');
     }
 
-    /*
-     *  Parse id_token
-     * */
-    protected function parse_token($launch)
+    /**
+     *  Get categories for a given course from canvas api
+     * @param $settings
+     * @return array
+     * @throws GuzzleException
+     */
+    private function get_categories($settings)
```

Comment:
There is already a method for getting groupcategories here: https://github.com/matematikk-mooc/kpas-api/blob/dcdbd6c415a7238c3302bb8e51128fdc2abba201/app/Services/CanvasService.php

Can you use that instead of creating duplicate code here?